pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -173,52 +209,47 @@ func (w *Watcher) Start() error {
 		return err
 	}
 
-	watcher, err := fsnotify.NewWatcher()
+	fsWatcher, err := fsnotify.NewWatcher()
 	if err != nil {
-		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
 	}
+	w.fsWatcher = fsWatcher
 
-	if err := watcher.Add(w.path); err != nil {
-		watcher.Close()
-		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	if err := fsWatcher.Add(w.path); err != nil {
+		fsWatcher.Close()
+		return fmt.Errorf("failed to watch %s, err: %v", w.path, err)
 	}
 
-	w.watcher = watcher
-
-	if err := w.traversePluginDir(); err != nil {
-		watcher.Close()
+	if err := w.traversePluginDir(w.path); err != nil {
+		fsWatcher.Close()
 		return fmt.Errorf("failed to traverse plugin socket path, err: %v", err)
 	}
 
 	w.wg.Add(1)
-	go func(watcher *fsnotify.Watcher) {
+	go func(fsWatcher *fsnotify.Watcher) {
 		defer w.wg.Done()
 		for {
 			select {
-			case event := <-watcher.Events:
-				if event.Op&fsnotify.Create == fsnotify.Create {
-					go func(eventName string) {
-						err := w.registerPlugin(eventName)
-						if err != nil {
-							glog.Errorf("Plugin %s registration failed with error: %v", eventName, err)
-						}
-					}(event.Name)
-				}
+			case event := <-fsWatcher.Events:
+				go func() {
+					err := w.handlefsnotifyEvent(event)
+					if err != nil {
+						glog.Errorf("error %s when handle event: %s", err, event)
+					}
+				}()
 				continue
-			case err := <-watcher.Errors:
-				//TODO: Handle errors by taking corrective measures
+			case err := <-fsWatcher.Errors:
 				if err != nil {
-					glog.Errorf("Watcher received error: %v", err)
+					glog.Errorf("fsWatcher received error: %v", err)
 				}
 				continue
-
 			case <-w.stopCh:
-				watcher.Close()
+				fsWatcher.Close()
 				break
```

Comment:
Hi @figo,

Be careful when using `break` and `continue` inside of `select` statements. For example, the `break` statement in `case <-w.stopCh:` doesn't break out of the loop but rather out of the `select` statement. See this Google group [post](https://groups.google.com/forum/#!topic/golang-nuts/s9w2gAlrpjk) for more information.